File: 118.pascals-triangle.cpp
/*
 * @lc app=leetcode id=118 lang=cpp
 *
 * [118] Pascal's Triangle
 */

// @lc code=start
class Solution {
 public:
  vector<vector<int>> generate(int numRows) {
    vector<vector<int>> ret;
    for (int i = 0; i < numRows; i++) {
      vector<int> row;
      for (int j = 0; j < i + 1; j++) {
        if ((j == 0) || (j == i)) {
          row.push_back(1);
        } else {
          row.push_back(ret[i - 1][j - 1] + ret[i - 1][j]);
        }
      }
      ret.push_back(row);
    }
    return ret;
  }
};
// @lc code=end

File: 118.pascals-triangle.rs
/*
 * @lc app=leetcode id=118 lang=rust
 *
 * [118] Pascal's Triangle
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[macro_export]
macro_rules! list_node {
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let mut node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        let num_rows: usize = num_rows as usize;
        let mut ret = vec![vec![1]];
        for i in 1..num_rows {
            let mut row = Vec::new();
            row.resize(i + 1, 1);
            for j in 1..i {
                row[j] = ret[i - 1][j - 1] + ret[i - 1][j];
            }
            ret.push(row)
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::generate(5),
        vec![
            vec![1],
            vec![1, 1],
            vec![1, 2, 1],
            vec![1, 3, 3, 1],
            vec![1, 4, 6, 4, 1]
        ]
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::generate(1), vec![vec![1]]);
}
